src/Storages/MergeTree/MergeTask.cpp

```diff
@@ -889,6 +901,14 @@ void MergeTask::ExecuteAndFinalizeHorizontalPart::createMergedStream()
     global_ctx->horizontal_stage_progress = std::make_unique<MergeStageProgress>(
         ctx->column_sizes ? ctx->column_sizes->keyColumnsWeight() : 1.0);
 
+    if (data_settings->allow_experimental_block_number_column
+        && !ctx->need_remove_expired_values && global_ctx->metadata_snapshot->getProjections().empty()
```

Comment:
Why we shouldn't have any projections? What if someone will add projection after `block_number` already exists?

Reply:
works with projection, updated check